### DISCUSSION TITLE
Using PInvoke to add new method to Solver.cs

### DISCUSSION THREAD

**Discussion by zbatool** [google/or-tools#2746] [Feature requests]
Hi, I need a C# method to load and solve mps files with the linear solver and am therefore trying to add some code to Solver.cs based on the solve.cc C++ example that reads and solves an mps file. Based on the other methods in Solver.cs I am trying to use PInvoke to call underlying C++ code e.g. how CreateSolver uses "operations_research_linear_solverPINVOKE.Solver_CreateSolver(solver_id)". 
I am trying to invoke the MPModelProto class that is defined in linear_solver.pb.h in namespace operations_research. I wrote the line "operations_researchPINVOKE.Proto_MPModelProto;" but got the error "The name 'operations_researchPINVOKE' does not exist in the current context [C:\or-tools-master-new\or-tools\temp_dotnet\Google.OrTools\Google.OrTools.csproj]".

How do I use PInvoke for other namespaces other than operations_research_linear_solver which is already being used in the code? Do I have to create some files elsewhere or change some dependencies in a csproj file somewhere?

**Comment by Mizux** [comment]
If you see some PINVOKE, it means the C++ code has not been correctly wrapped into .Net by SWIG

C++ interface
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/linear_solver/model_exporter.h#L100-L104

Rebinded to get rid of the `absl::StatusOr<std::string>` return type (ed would be a nightmare to wrap absl template in .Net)
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/linear_solver/model_exporter_swig_helper.h#L31-L35

And the SWIG code which will use it
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/linear_solver/csharp/linear_solver.i#L166-L188

**Reply by zbatool** [reply]
Hi, I'm not sure how PInvoke declarations are used in this case then as model_exporter.h, model_exporter_swig_helper.h and linear_solver.i do not contain many of the other functions used in Solver.cs. For context I am trying to do the following: add a new method for reading and solving mps files in or-tools\ortools\linear_solver\linear_solver.h that I can then invoke potentially using Swig and PInvoke in C# \or-tools\ortools\gen\ortools\linear_solver\Solver.cs. I need to understand what files I need to make changes in to be able to call it through something like "operations_research_linear_solverPINVOKE.Solver_SolveMpsFile" in Solver.cs.

For example CreateSolver method in Solver.cs in \or-tools\ortools\gen\ortools\linear_solver there is a line "global::System.IntPtr cPtr = operations_research_linear_solverPINVOKE.Solver_CreateSolver(solver_id);" so I want to do a similar invocation of code for a C++ method that I create

**Reply by zbatool** [reply]
If i just add a test method in or-tools\ortools\linear_solver\linear_solver.h, clean-make and rebuild the project and then try and reference it in Solver.cs I get 'operations_research_linear_solverPINVOKE' does not contain a definition for 'Solver_SolveMpsFile' [C:\or-tools-master-new\or-tools\temp_dotnet\Google.OrTools\Google.OrTools.csproj]

**Comment by lperron** [comment]
I have this basic question. Once loaded, you cannot in inspect the solution easily. So what is the workflow here ?

**Reply by zbatool** [reply]
Hi, I actually need the functionality to read and solve mps files in C# (similar to the functionality in example solve.cc) so I am trying to create a method as I discussed in my reply above to do this. I am thinking of doing either of two things:
1) writing a C++ method in or-tools\ortools\linear_solver\linear_solver.h to do the same and then PInvoke-ing it in  \or-tools\ortools\gen\ortools\linear_solver\Solver.cs or
2) using PInvoke or some other way to do all the steps outlined in solve.cc but for C# in a new method in  \or-tools\ortools\gen\ortools\linear_solver\Solver.cs
but am having trouble using PInvoke for new code I write and am wondering what steps I need to follow if I need to reference new C++ code in \or-tools\ortools\gen\ortools\linear_solver\Solver.cs

Hope that clarified my question
